src/libraries/System.Text.Json/src/System/Text/Json/ThrowHelper.Serialization.cs

```diff
@@ -94,9 +94,15 @@ public static void ThrowJsonException(string? message = null)
 
         [DoesNotReturn]
         [MethodImpl(MethodImplOptions.NoInlining)]
-        public static void ThrowInvalidOperationException_CannotSerializeOpenGeneric(Type type)
+        public static void ThrowInvalidOperationException_CannotSerializeInvalidType(Type type, Type? parentClassType, PropertyInfo? propertyInfo)
         {
-            throw new InvalidOperationException(SR.Format(SR.CannotSerializeOpenGeneric, type));
+            if (parentClassType == null)
+            {
+                Debug.Assert(propertyInfo == null);
+                throw new InvalidOperationException(SR.Format(SR.CannotSerializeInvalidType, type));
+            }
+
+            throw new InvalidOperationException(SR.Format(SR.CannotSerializeInvalidMember, type, propertyInfo!.Name, parentClassType));
```

Comment:
If I understood your comment correctly (see https://github.com/dotnet/runtime/pull/34777#discussion_r412679793), then this can null ref when `propertyInfo` is null. We shouldn't be using the null-forgiving `!` in this case to work around the compiler warning, but rather handle the case when `propertyInfo` is null correctly. This is what I was referring to earlier.

Reply:
Thanks. I've added the assert - it's useful for guarding against future breaks as the code evolves, and it better conveys the intent.